src/cdk-experimental/selection/selection.ts

```diff
@@ -0,0 +1,226 @@
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {coerceBooleanProperty} from '@angular/cdk/coercion';
+import {CollectionViewer, DataSource, isDataSource, ListRange} from '@angular/cdk/collections';
+import {
+  AfterContentChecked,
+  Directive,
+  EventEmitter,
+  Input,
+  OnDestroy,
+  OnInit,
+  Output,
+  TrackByFunction
+} from '@angular/core';
+import {Observable, of as observableOf, ReplaySubject, Subscription} from 'rxjs';
+import {takeUntil} from 'rxjs/operators';
+
+import {SelectableWithIndex, SelectionChange, SelectionSet} from './selection-set';
+
+/**
+ * Manages the selection states of the items and provides methods to check and update the selection
+ * states.
+ * It must be applied to the parent element if `cdkSelectionToggle`, `cdkSelectAll`,
+ * `cdkRowSelection` and `cdkSelectionColumn` are applied.
+ */
+@Directive({
+  selector: '[cdkSelection]',
+  exportAs: 'cdkSelection',
+})
+export class CdkSelection<T> implements OnInit, AfterContentChecked, CollectionViewer, OnDestroy {
+  viewChange: Observable<ListRange>;
+
+  @Input()
+  get dataSource(): TableDataSource<T> {
+    return this._dataSource;
+  }
+  set dataSource(dataSource: TableDataSource<T>) {
+    if (this._dataSource !== dataSource) {
+      this._switchDataSource(dataSource);
+    }
+  }
+  private _dataSource: TableDataSource<T>;
+
+  @Input()
+  get trackBy(): TrackByFunction<T> {
+    return this._trackByFn;
+  }
+  set trackBy(fn: TrackByFunction<T>) {
+    this._trackByFn = fn;
+  }
+  private _trackByFn: TrackByFunction<T>;
+
+  /** Whether to support multiple selection */
+  @Input()
+  get cdkSelectionMultiple(): boolean {
+    return this._multiple;
+  }
+  set cdkSelectionMultiple(multiple: boolean) {
+    this._multiple = coerceBooleanProperty(multiple);
+  }
+  private _multiple: boolean;
+
+  /** Emits when selection changes. */
+  @Output() cdkSelectionChange = new EventEmitter<SelectionChange<T>>();
+
+  /** Latest data provided by the data source. */
+  private _data: T[]|readonly T[];
+
+  /** Subscription that listens for the data provided by the data source.  */
+  private _renderChangeSubscription: Subscription|null;
+
+  private _destroyed$ = new ReplaySubject<void>(1);
+
+  private _selection: SelectionSet<T>;
+
+  private _switchDataSource(dataSource: TableDataSource<T>) {
+    this._data = [];
+
+    if (isDataSource(this._dataSource)) {
+      this._dataSource.disconnect(this);
+    }
+
+    if (this._renderChangeSubscription) {
+      this._renderChangeSubscription.unsubscribe();
+      this._renderChangeSubscription = null;
+    }
+
+    this._dataSource = dataSource;
+  }
+
+  private _observeRenderChanges() {
+    if (!this._dataSource) {
+      return;
+    }
+
+    let dataStream: Observable<T[]|ReadonlyArray<T>>|undefined;
+
+    if (isDataSource(this._dataSource)) {
```

Comment:
Could you add a TODO here to move this logic to a shared function in `cdk/collections`?